### PR TITLE
Initial commit

### PR DESCRIPTION
**Pull request by VadimDomantsevich**

### CODE REVIEW

File: Lab2/src/BLL/BLL.csproj [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+  <PropertyGroup>
+    <TargetFramework>netstandard2.0</TargetFramework>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="Autofac" Version="5.1.2" />
+    <PackageReference Include="AutoMapper" Version="9.0.0" />
+    <PackageReference Include="AutoMapper.Extensions.Microsoft.DependencyInjection" Version="7.0.0" />
+    <PackageReference Include="Microsoft.EntityFrameworkCore" Version="3.1.3" />
```
Comment (alexey62soft) [inline Lab2/src/BLL/BLL.csproj:11]: EF libs should be in DAL only

File: Lab2/src/BLL/DI/DIConfigurator.cs [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+using System.Reflection;
+
+namespace BLL.DI
+{
+    public static class DIConfigurator
+    {
+        public static ServiceProvider Configure(IConfigurationRoot config)
+        {
+            var businessAssembly = Assembly.Load("BLL");
+            var consoleAssembly = Assembly.Load("UI");
```
Comment (alexey62soft) [inline Lab2/src/BLL/DI/DIConfigurator.cs]: It shouldn't be in BLL (incorrect project inheritance hierarchy)

File: Lab2/src/BLL/DI/DIConfigurator.cs [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+using Microsoft.EntityFrameworkCore;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.DependencyInjection;
+using System.Reflection;
+
+namespace BLL.DI
+{
+    public static class DIConfigurator
+    {
+        public static ServiceProvider Configure(IConfigurationRoot config)
```
Comment (alexey62soft) [inline Lab2/src/BLL/DI/DIConfigurator.cs]: Better to move it to UI project

File: Lab2/src/BLL/Interfaces/IGroupServices.cs [hunk 4]
```diff
@@ -0,0 +1,12 @@
+﻿using BLL.Models;
+using System.Collections.Generic;
+
+namespace BLL.Interfaces
+{
+    public interface IGroupServices : IServices<Group>
+    {
+        IEnumerable<Group> GetGroupsBySpecialtyId(int specialtyId);
```
Comment (alexey62soft) [inline Lab2/src/BLL/Interfaces/IGroupServices.cs]: All methods should be awaitable.

File: Lab2/src/BLL/Interfaces/IServices.cs [hunk 5]
```diff
@@ -0,0 +1,17 @@
+﻿using System.Collections.Generic;
+
+namespace BLL.Interfaces
+{
+    public interface IServices<T>
+    {
+        void Create(T item);
```
Comment (alexey62soft) [inline Lab2/src/BLL/Interfaces/IServices.cs]: Awaitable.

File: Lab2/src/BLL/Models/Student.cs [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+        public int Id { get; set; }
+
+        public string FullName { get; set; }
+
+        public string RecordNumber { get; set; }
+
+        public int GroupId { get; set; }
+
+        public string PhoneNumber { get; set; }
+
```
Comment (alexey62soft) [inline Lab2/src/BLL/Models/Student.cs:14]: useless space

File: Lab2/src/BLL/BLL.csproj [hunk 7]
```diff
@@ -0,0 +1,20 @@
+<Project Sdk="Microsoft.NET.Sdk">
```
Comment (alexey62soft) [inline Lab2/src/BLL/BLL.csproj:1]: remove `src` folder

File: Lab2/src/BLL/Interfaces/IStudentServices.cs [hunk 8]
```diff
@@ -0,0 +1,14 @@
+﻿using System.Collections.Generic;
+using BLL.Models;
+
+namespace BLL.Interfaces
+{
+    public interface IStudentServices : IServices<Student>
```
Comment (alexey62soft) [inline Lab2/src/BLL/Interfaces/IStudentServices.cs]: IStudentsService

File: Lab2/src/BLL/Interfaces/IStatementServices.cs [hunk 9]
```diff
@@ -0,0 +1,14 @@
+﻿using System.Collections.Generic;
+using BLL.Models;
+
+namespace BLL.Interfaces
+{
+    public interface IStatementServices : IServices<Statement>
```
Comment (alexey62soft) [inline Lab2/src/BLL/Interfaces/IStatementServices.cs]: IStatementsService

File: Lab2/src/BLL/Interfaces/ISubjectServices.cs [hunk 10]
```diff
@@ -0,0 +1,9 @@
+﻿using BLL.Models;
+
+namespace BLL.Interfaces
+{
+    interface ISubjectServices : IServices<Subject>
```
Comment (alexey62soft) [inline Lab2/src/BLL/Interfaces/ISubjectServices.cs]: ISubjectsService

File: Lab2/src/BLL/Interfaces/IServices.cs [hunk 11]
```diff
@@ -0,0 +1,17 @@
+﻿using System.Collections.Generic;
+
+namespace BLL.Interfaces
+{
+    public interface IServices<T>
```
Comment (alexey62soft) [inline Lab2/src/BLL/Interfaces/IServices.cs]: IService

File: Lab2/src/BLL/Services/StatementServices.cs [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+using BLL.Interfaces;
+using DAL.Interfaces;
+using DAL.DTO;
+using System.Collections.Generic;
+using System.Linq;
+using AutoMapper;
+
+namespace BLL.Services
+{
+    public class StatementServices : IStatementServices
```
Comment (alexey62soft) [inline Lab2/src/BLL/Services/StatementServices.cs:12]: StatementsService (refactor all services names)

File: Lab2/src/DAL/DALClassDiagram.cd [hunk 13]
```diff
@@ -0,0 +1,2 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
```
Comment (alexey62soft) [inline Lab2/src/DAL/DALClassDiagram.cd]: What? Empty?

File: Lab2/src/BLL/Models/TypeOfSertification.cs [hunk 14]
```diff
@@ -0,0 +1,9 @@
+﻿namespace BLL.Models
+{
+    public enum TypeOfSertification
```
Comment (alexey62soft) [inline Lab2/src/BLL/Models/TypeOfSertification.cs]: Use enum from DAL

File: Lab2/src/DAL/EntityFramework/DBContext.cs [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+
+        public DbSet<SubjectDTO> Subjects { get; set; }
+
+        protected override void OnModelCreating(ModelBuilder modelBuilder)
+        {
+            modelBuilder.Entity<GroupDTO>()
+                .HasOne<SpecialtyDTO>()
+                .WithMany()
+                .HasForeignKey(group => group.SpecialtyId)
+                .OnDelete(DeleteBehavior.NoAction);
```
Comment (alexey62soft) [inline Lab2/src/DAL/EntityFramework/DBContext.cs:36]: No action?
Make configuration with dependencies removing.

File: Lab2/src/DAL/Interfaces/IRepository.cs [hunk 16]
```diff
@@ -0,0 +1,18 @@
+﻿using System.Linq;
+
+namespace DAL.Interfaces
+{
+    public interface IRepository<T>
+        where T : class
+    {
+        IQueryable<T> GetAll();
```
Comment (alexey62soft) [inline Lab2/src/DAL/Interfaces/IRepository.cs:8]: All methods should be awaitable

File: Lab2/src/DAL/EntityFramework/Repository.cs [hunk 17]
```diff
@@ -0,0 +1,51 @@
+﻿using DAL.Interfaces;
+using Microsoft.EntityFrameworkCore;
+using System.Linq;
+
+namespace DAL.EntityFramework
+{
+    public class Repository<T> : IRepository<T>
```
Comment (alexey62soft) [inline Lab2/src/DAL/EntityFramework/Repository.cs]: GenericRepository<T>

File: Lab2/src/UI/Interfaces/ICrudConsoleService.cs [hunk 18]
```diff
@@ -0,0 +1,13 @@
+﻿namespace UI.Interfaces
+{
+    public interface ICrudConsoleService<T>
+    {
+        void Create();
```
Comment (alexey62soft) [inline Lab2/src/UI/Interfaces/ICrudConsoleService.cs:5]: Awaitable

File: Lab2/src/UI/Program.cs [hunk 19]
```diff
@@ -0,0 +1,26 @@
...
+        {
+            var builder = new ConfigurationBuilder();
+            builder.SetBasePath(Directory.GetCurrentDirectory());
+            builder.AddJsonFile(_configFileName);
+            var config = builder.Build();
+
+            var services = DIConfigurator.Configure(config);
+
+            var mainMenu = services.GetService<MainMenuBaseConsoleService>();
+            mainMenu.StartLoop();
```
Comment (alexey62soft) [inline Lab2/src/UI/Program.cs:23]: awaitable

PS we can create async Task Main()

File: Lab2/src/UI/Interfaces/IConsoleService.cs [hunk 20]
```diff
@@ -0,0 +1,9 @@
+﻿namespace UI.Interfaces
+{
+    public interface IConsoleService
+    {
+        void PrintMenu();
+
+        void StartLoop();
```
Comment (alexey62soft) [inline Lab2/src/UI/Interfaces/IConsoleService.cs:7]: Awaitable

Comment (alexey62soft) [reply Lab2/src/UI/Interfaces/IConsoleService.cs:7]: only this I guess

File: Lab2/src/UI/UIClassDiagram.cd [hunk 21]
```diff
@@ -0,0 +1,2 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
```
Comment (alexey62soft) [inline Lab2/src/UI/UIClassDiagram.cd]: What? Empty?

File: Lab2/src/UI/View/CrudConsoleService/GroupCrudConsoleService.cs [hunk 22]
```diff
@@ -0,0 +1,61 @@
+﻿using BLL.Models;
+using BLL.Services;
+using System;
+using UI.Interfaces;
+
+namespace UI.View.CrudConsoleService
+{
+    public class GroupCrudConsoleService : ICrudConsoleService<Group>
```
Comment (alexey62soft) [inline Lab2/src/UI/View/CrudConsoleService/GroupCrudConsoleService.cs]: GroupsCrudConsoleService (refactor all names for all CrudConsole services)

File: Lab2/src/UI/View/ConsolePrinter.cs [hunk 23]
```diff
@@ -0,0 +1,13 @@
+﻿using ConsoleTables;
+using System.Collections.Generic;
+
+namespace UI.View
```
Comment (alexey62soft) [inline Lab2/src/UI/View/ConsolePrinter.cs]: View?
Helpers better folder for this.

And make ConsoleHelpers class for this.

File: Lab2/src/UI/View/CrudConsoleService/GroupCrudConsoleService.cs [hunk 24]
```diff
@@ -0,0 +1,61 @@
+﻿using BLL.Models;
+using BLL.Services;
+using System;
+using UI.Interfaces;
+
+namespace UI.View.CrudConsoleService
```
Comment (alexey62soft) [inline Lab2/src/UI/View/CrudConsoleService/GroupCrudConsoleService.cs]: View?
Services - better folder and withoout additional `CrudConsoleService` folder

File: Lab2/src/UI/View/MainConsoleService/GroupBaseConsoleService.cs [hunk 25]
```diff
@@ -0,0 +1,68 @@
+﻿using BLL.Infrastructure;
+using BLL.Models;
+using System;
+using UI.Interfaces;
+using UI.View.PrintConsoleService;
+
+namespace UI.View.MainConsoleService
```
Comment (alexey62soft) [inline Lab2/src/UI/View/MainConsoleService/GroupBaseConsoleService.cs]: Same as before

File: Lab2/src/UI/View/MainConsoleService/GroupBaseConsoleService.cs [hunk 26]
```diff
@@ -0,0 +1,68 @@
...
+{
+    public class GroupBaseConsoleService : IConsoleService
```
Comment (alexey62soft) [inline Lab2/src/UI/View/MainConsoleService/GroupBaseConsoleService.cs]: Same situation with names

File: Lab2/src/UI/Interfaces/IPrintConsoleService.cs [hunk 27]
```diff
@@ -0,0 +1,7 @@
+﻿namespace UI.Interfaces
+{
+    public interface IPrintConsoleService
+    {
+        void PrintAll();
```
Comment (alexey62soft) [inline Lab2/src/UI/Interfaces/IPrintConsoleService.cs:5]: awaitable

File: Lab2/src/UI/View/PrintConsoleService/GroupPrintConsoleService.cs [hunk 28]
```diff
@@ -0,0 +1,33 @@
+﻿using BLL.Services;
+using System.Linq;
+using UI.Interfaces;
+using UI.ViewModels;
+
+namespace UI.View.PrintConsoleService
+{
+    public class GroupPrintConsoleService : IPrintConsoleService
```
Comment (alexey62soft) [inline Lab2/src/UI/View/PrintConsoleService/GroupPrintConsoleService.cs]: Same situation with names as above

File: Lab2/src/UI/View/PrintConsoleService/GroupPrintConsoleService.cs [hunk 29]
```diff
@@ -0,0 +1,33 @@
+﻿using BLL.Services;
+using System.Linq;
+using UI.Interfaces;
+using UI.ViewModels;
+
+namespace UI.View.PrintConsoleService
```
Comment (alexey62soft) [inline Lab2/src/UI/View/PrintConsoleService/GroupPrintConsoleService.cs]: Same as above

**Review by alexey62soft — CHANGES_REQUESTED** [review]
A lot of issues that were discussed before.
Bad work!

File: Lab2/src/BLL/BLL.csproj (hunk 1 above)
Comment (VadimDomantsevich) [reply Lab2/src/BLL/BLL.csproj:11]: Fixed

File: Lab2/src/BLL/DI/DIConfigurator.cs (hunk 2 above)
Comment (VadimDomantsevich) [reply Lab2/src/BLL/DI/DIConfigurator.cs]: Fixed

File: Lab2/src/BLL/DI/DIConfigurator.cs (hunk 3 above)
Comment (VadimDomantsevich) [reply Lab2/src/BLL/DI/DIConfigurator.cs]: Fixed

File: Lab2/src/BLL/Models/Student.cs (hunk 6 above)
Comment (VadimDomantsevich) [reply Lab2/src/BLL/Models/Student.cs:14]: Fixed

File: Lab2/src/BLL/Interfaces/IStudentServices.cs (hunk 8 above)
Comment (VadimDomantsevich) [reply Lab2/src/BLL/Interfaces/IStudentServices.cs]: Fixed

File: Lab2/src/BLL/Interfaces/IStatementServices.cs (hunk 9 above)
Comment (VadimDomantsevich) [reply Lab2/src/BLL/Interfaces/IStatementServices.cs]: Fixed

File: Lab2/src/BLL/Interfaces/ISubjectServices.cs (hunk 10 above)
Comment (VadimDomantsevich) [reply Lab2/src/BLL/Interfaces/ISubjectServices.cs]: Fixed

File: Lab2/src/BLL/Interfaces/IServices.cs (hunk 11 above)
Comment (VadimDomantsevich) [reply Lab2/src/BLL/Interfaces/IServices.cs]: Fixed

File: Lab2/src/BLL/Services/StatementServices.cs (hunk 12 above)
Comment (VadimDomantsevich) [reply Lab2/src/BLL/Services/StatementServices.cs:12]: Fixed

File: Lab2/src/BLL/Models/TypeOfSertification.cs (hunk 14 above)
Comment (VadimDomantsevich) [reply Lab2/src/BLL/Models/TypeOfSertification.cs]: Fixed

File: Lab2/src/DAL/EntityFramework/DBContext.cs (hunk 15 above)
Comment (VadimDomantsevich) [reply Lab2/src/DAL/EntityFramework/DBContext.cs:36]: Fixed

File: Lab2/src/DAL/Interfaces/IRepository.cs (hunk 16 above)
Comment (VadimDomantsevich) [reply Lab2/src/DAL/Interfaces/IRepository.cs:8]: Fixed

File: Lab2/src/DAL/EntityFramework/Repository.cs (hunk 17 above)
Comment (VadimDomantsevich) [reply Lab2/src/DAL/EntityFramework/Repository.cs]: Fixed

File: Lab2/src/BLL/Interfaces/IGroupServices.cs (hunk 4 above)
Comment (VadimDomantsevich) [reply Lab2/src/BLL/Interfaces/IGroupServices.cs]: Fixed

File: Lab2/src/BLL/Interfaces/IServices.cs (hunk 5 above)
Comment (VadimDomantsevich) [reply Lab2/src/BLL/Interfaces/IServices.cs]: Fixed

File: Lab2/src/UI/Interfaces/ICrudConsoleService.cs (hunk 18 above)
Comment (VadimDomantsevich) [reply Lab2/src/UI/Interfaces/ICrudConsoleService.cs:5]: Fixed

File: Lab2/src/UI/Program.cs (hunk 19 above)
Comment (VadimDomantsevich) [reply Lab2/src/UI/Program.cs:23]: Fixed

File: Lab2/src/UI/Interfaces/IConsoleService.cs (hunk 20 above)
Comment (VadimDomantsevich) [reply Lab2/src/UI/Interfaces/IConsoleService.cs:7]: Fixed

File: Lab2/src/UI/View/CrudConsoleService/GroupCrudConsoleService.cs (hunk 22 above)
Comment (VadimDomantsevich) [reply Lab2/src/UI/View/CrudConsoleService/GroupCrudConsoleService.cs]: Fixed

File: Lab2/src/UI/View/ConsolePrinter.cs (hunk 23 above)
Comment (VadimDomantsevich) [reply Lab2/src/UI/View/ConsolePrinter.cs]: Fixed

File: Lab2/src/UI/View/CrudConsoleService/GroupCrudConsoleService.cs (hunk 24 above)
Comment (VadimDomantsevich) [reply Lab2/src/UI/View/CrudConsoleService/GroupCrudConsoleService.cs]: Fixed

File: Lab2/src/UI/View/MainConsoleService/GroupBaseConsoleService.cs (hunk 25 above)
Comment (VadimDomantsevich) [reply Lab2/src/UI/View/MainConsoleService/GroupBaseConsoleService.cs]: Fixed

File: Lab2/src/UI/View/MainConsoleService/GroupBaseConsoleService.cs (hunk 26 above)
Comment (VadimDomantsevich) [reply Lab2/src/UI/View/MainConsoleService/GroupBaseConsoleService.cs]: Fixed

File: Lab2/src/UI/Interfaces/IPrintConsoleService.cs (hunk 27 above)
Comment (VadimDomantsevich) [reply Lab2/src/UI/Interfaces/IPrintConsoleService.cs:5]: Fixed

File: Lab2/src/UI/View/PrintConsoleService/GroupPrintConsoleService.cs (hunk 28 above)
Comment (VadimDomantsevich) [reply Lab2/src/UI/View/PrintConsoleService/GroupPrintConsoleService.cs]: Fixed

File: Lab2/src/UI/View/PrintConsoleService/GroupPrintConsoleService.cs (hunk 29 above)
Comment (VadimDomantsevich) [reply Lab2/src/UI/View/PrintConsoleService/GroupPrintConsoleService.cs]: Fixed

File: Lab2/src/UI/UIClassDiagram.cd (hunk 21 above)
Comment (VadimDomantsevich) [reply Lab2/src/UI/UIClassDiagram.cd]: Fixed

File: Lab2/src/DAL/DALClassDiagram.cd (hunk 13 above)
Comment (VadimDomantsevich) [reply Lab2/src/DAL/DALClassDiagram.cd]: Fixed

**Pull request closed without merging**